Add Resolution Presets for the embedded game window

Kazox61 opened the pull request:
Partial implementation of [#13287](https://github.com/godotengine/godot-proposals/issues/13287).

This adds a dictionary to the editor settings to define new resolution presetes.
In the embeded game window you can select one of them to change the resolution.


https://github.com/user-attachments/assets/d30cf8d6-f1b9-4f1b-a730-497e8f8849eb

AThousandShips commented on editor/run/game_view_plugin.cpp:
```diff
@@ -501,6 +501,19 @@ void GameView::_embedded_process_focused() {
+	PopupMenu *menu = resolution_options_menu->get_popup();
+	menu->clear();
+
+	menu->add_radio_check_item("Default (" + itos(EditorRun::get_window_placement().size.x) + "x" + itos(EditorRun::get_window_placement().size.y) + ")", 0);
```
```suggestion
	menu->add_radio_check_item(vformat(TTR("Default (%dx%d)"), EditorRun::get_window_placement().size.x, EditorRun::get_window_placement().size.y), 0);
```

Nintorch commented:
It seems that you've accidentally rebased incorrectly, thus a lot of commits unrelated to this PR have been added 😅

Kazox61 commented:
Yes🤣

Kazox61 committed:
Add Resolution Presets for the embedded game window

Calinou commented:
@Kazox61 Can you look into rebasing this PR to fix merge conflicts? I'd like to review this for the 4.8 cycle.

Given the video in OP, I can give some early feedback still:

- The dictionary keys should be String, not StringName. StringName should be reserved for strings that use very frequent lookups to benefit from string interning. Otherwise, it's not worth using  StringName.

- The UX for editing settings isn't ideal, as the dictionary isn't typed and you can add invalid types as keys/values. Is it possible to use a typed dictionary for both new project settings?

- Landscape and portrait hints for safe areas should be `.landscape` and `.portrait`. Like in [feature tags](https://docs.godotengine.org/en/latest/tutorials/export/feature_tags.html), we tend to avoid abbreviations for readability.
  - Alternatively, is it possible to assume landscape mode for safe area definitions, and automatically rotate the area to match portrait mode when the viewport height is greater than the width? This would reduce the work needed to add a new device.

- We should provide a few Android devices in the list as defaults. See the list in Chromium 150 (although we should use more recent devices):

<img width="183" height="411" alt="image" src="https://github.com/user-attachments/assets/92a61f9c-d867-49a3-a42e-105ea4a2f460" />

Kazox61 commented:
@Calinou I rebased and added all your suggestions.

Calinou reviewed (commented):
Tested locally, it works as expected. I think this is a great start, but some changes will be needed to make sure this better matches user expectations.

Some feedback:

- When opening the editor, there is no resolution preset initially checked:

<img width="376" height="255" alt="Image" src="https://github.com/user-attachments/assets/44f2b2c7-b90e-47e6-b1de-1652ceb6ba3e" />

Instead, the Default preset should be checked by default.

- With this PR, overrides set in the project settings are no longer obeyed when the resolution preset is set to Default:

<img width="605" height="104" alt="Image" src="https://github.com/user-attachments/assets/c7ca3669-318d-4510-ad0e-ce1d6160c35e" />

Instead, it always uses the base window width/height defined in the project settings. This should be reverted to the previous behavior to match what happens when game embedding is disabled. The Default resolution text should also display the effective resolution that takes the override into account (if any).

- The Samsung Galaxy S24 and S25 series presets have the exact same resolution. I suggest removing all Galaxy S24 entries to shorten the list.
  - The same goes for the OnePlus 12 and 13; you can remove the OnePlus 12 from the presets.

- I noticed Android devices don't have a safe area defined in their presets. Is that intended?

- The orientation project setting is incorrectly followed. All settings other than **Landscape** will use portrait mode for resolution presets:

<img width="968" height="242" alt="Image" src="https://github.com/user-attachments/assets/7ecd0072-0c66-41cf-8b30-c28de35f6b87" />

Instead, **Landscape**, **Reverse Landscape** and **Sensor Landscape** should use a landscape resolution.

- Resolution presets have no effect if the **Stretch to Fit** scale mode is selected in game embedding.
  - To allow for a smoother UX, I suggest automatically swapping back to **Keep Aspect Ratio** when selecting a preset. Also, when choosing **Stretch to Fit**, the previously selected preset should be unchecked in the list.

- When choosing a preset, it enforces a maximum resolution and aspect ratio. It does not allow the viewport to render at a higher resolution than what's shown on screen, even when using the `viewport` stretch mode. This can result in different visuals than what you'll see on the target device due to the pixel count being different. Some shaders have resolution-dependent behavior, and of course, using the `disabled` stretch mode will also result in resolution-dependent visuals.

<img width="494" height="261" alt="Image" src="https://github.com/user-attachments/assets/dd4fd4ae-03dd-4e16-9545-a10d40892db1" />

Kazox61 commented:
> When opening the editor, there is no resolution preset initially checked.
Instead, the Default preset should be checked by default.

Fixed

> With this PR, overrides set in the project settings are no longer obeyed when the resolution preset is set to Default
Instead, it always uses the base window width/height defined in the project settings. This should be reverted to the previous behavior to match what happens when game embedding is disabled. The Default resolution text should also display the effective resolution that takes the override into account (if any).

Should be fixed

> The Samsung Galaxy S24 and S25 series presets have the exact same resolution. I suggest removing all Galaxy S24 entries to shorten the list.
The same goes for the OnePlus 12 and 13; you can remove the OnePlus 12 from the presets.

Done 

> I noticed Android devices don't have a safe area defined in their presets. Is that intended?

Yes it is. Android doesn't have standardized safe area insets. The status bar and gesture bar heights vary by OEM, Android version, and user settings. Maybe we could use some default values there, but I don't know much about it so left them out for now.

> The orientation project setting is incorrectly followed. All settings other than Landscape will use portrait mode for resolution presets. 
Instead, Landscape, Reverse Landscape and Sensor Landscape should use a landscape resolution.

Fixed

> Resolution presets have no effect if the Stretch to Fit scale mode is selected in game embedding.
To allow for a smoother UX, I suggest automatically swapping back to Keep Aspect Ratio when selecting a preset. Also, when choosing Stretch to Fit, the previously selected preset should be unchecked in the list.

So you mean when "Stretch to Fit" is active and you select a preset, the "Keep Aspect Ratio" should be selected. And when you select  "Stretch to Fit" the selected preset should get unselected?

> When choosing a preset, it enforces a maximum resolution and aspect ratio. It does not allow the viewport to render at a higher resolution than what's shown on screen, even when using the viewport stretch mode. This can result in different visuals than what you'll see on the target device due to the pixel count being different. Some shaders have resolution-dependent behavior, and of course, using the disabled stretch mode will also result in resolution-dependent visuals.

Any idea how we can handle this?

AThousandShips commented on editor/run/game_view_plugin.cpp:
```diff
@@ -1376,7 +1473,38 @@ void GameView::_update_arguments_for_instance(int p_idx, List<String> &r_argumen
+		case DisplayServerEnums::ScreenOrientation::SCREEN_LANDSCAPE:
+		case DisplayServerEnums::ScreenOrientation::SCREEN_REVERSE_LANDSCAPE:
+		case DisplayServerEnums::ScreenOrientation::SCREEN_SENSOR_LANDSCAPE:
+			resolution_name = StringName(String(resolution_name) + ".landscape");
```
```suggestion
			resolution_name = resolution_name + ".landscape";
```
Or just
```suggestion
			resolution_name += ".landscape";
```
For both, the value is a `String`:
```cpp
String resolution_name = resolution_presets.get_key_at_index(resolution_preset_index - 1);
```

Calinou commented:
> So you mean when "Stretch to Fit" is active and you select a preset, the "Keep Aspect Ratio" should be selected. And when you select "Stretch to Fit" the selected preset should get unselected?

Yes :slightly_smiling_face:

Kazox61 commented:
> So you mean when "Stretch to Fit" is active and you select a preset, the "Keep Aspect Ratio" should be selected. And when you select "Stretch to Fit" the selected preset should get unselected?

Added

Kazox61 commented:
One problem I noticed is that Resolutions are getting reordered when you reopen the Editor Settings. The saved index in metadata doesn't match anymore selected preset.

m4gr3d commented on doc/classes/EditorSettings.xml:
```diff
@@ -1410,6 +1410,14 @@
+		</member>
+		<member name="run/resolution_presets/safe_areas" type="Dictionary" setter="" getter="">
+			Safe areas for the resolution presets defined in [member run/resolution_presets/resolutions].
+			The key needs to match with the key in [member run/resolution_presets/resolutions] plus a ".l" for landscape or ".p" for portrait.
```
```suggestion
			The key needs to match with the key in [member run/resolution_presets/resolutions] plus a ".landscape" for landscape or ".portrait" for portrait.
```

m4gr3d commented on editor/settings/editor_settings.cpp:
```diff
@@ -1138,6 +1139,57 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("run/platforms/linuxbsd/prefer_wayland", false, true);
 	set_restart_if_changed("run/platforms/linuxbsd/prefer_wayland", true);
 
+	// Resolution presets
```
Can we visually group the resolution presets either per OS (Android, iOS) or per form factor (phone, tablets)?

This could be done by extending the metadata stored in the dictionary, and using that additional info to create submenus.

Kazox61 replied:
How should I extend the metadata in the dictionary?

m4gr3d replied:
One option could be to use a `PackedInt32Array` instead of `Size2i` to store the OS or form factor alongside the device dimensions.

m4gr3d commented on editor/settings/editor_settings.cpp:
```diff
@@ -1138,6 +1139,57 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("run/platforms/linuxbsd/prefer_wayland", false, true);
 	set_restart_if_changed("run/platforms/linuxbsd/prefer_wayland", true);
 
+	// Resolution presets
```
We should establish a criteria for expanding the resolution presets list, otherwise we'll have requests to include every single devices under the sun.

It could be that we set a max limit per OS (or form factor).

Calinou replied:
I would go for these criteria:

- Device must be from an established brand that is still on the market.
- No duplicate resolutions (and safe areas). If a device has the same resolution as another device already on the list, then it should be skipped, unless it has a different safe area.
  - If several devices match the same resolution/safe area combo, then the most recent device should be listed as long as it fits other criteria here.
  - If resolutions/safe areas differ, only list the two most recent devices of each lineup (e.g. iPhone 16/17).

There's still the open question of how we should go about updating the list of presets across minor releases. Should we remove old device presets eventually, or keep them in until Godot 5.0 is released? They can always be recreated as custom presets after all, so perhaps keeping the list tidy is more important from an usability perspective.

m4gr3d replied:
I think we should remove old device presets as we update the list across minor releases. We could include an `export` button to allow exporting the hardcoded list from an older release so it can be reused in a newer release that's missing some of the old entries.

m4gr3d commented on editor/settings/editor_settings.cpp:
```diff
@@ -1138,6 +1139,57 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("run/platforms/linuxbsd/prefer_wayland", false, true);
 	set_restart_if_changed("run/platforms/linuxbsd/prefer_wayland", true);
 
+	// Resolution presets
```
In the comment, can you add the source of the data, so it's easy for others to update using the same source when needed.

Kazox61 replied:
I got them from different sources

Calinou replied:
GSMArena lists screen resolutions, but not safe areas. I'm not sure what's the most reliable source for those.

m4gr3d reviewed (commented):
@syntaxerror247 We'll need to replicate this feature in the Android editor.

Calinou commented on editor/settings/editor_settings.cpp:
```diff
@@ -1138,6 +1139,57 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
+	default_resolutions[String("iPad Pro 13")] = Size2i(2064, 2752);
+	default_resolutions[String("Samsung Galaxy S25")] = Size2i(1080, 2340);
+	default_resolutions[String("Samsung Galaxy S25+")] = Size2i(1440, 3120);
+	default_resolutions[String("Samsung Galaxy S25 Ultra")] = Size2i(1440, 3120);
```
The S26 series has the same resolution as S25, might as well list that one instead (as per my above comment):

```suggestion
	default_resolutions[String("Samsung Galaxy S26")] = Size2i(1080, 2340);
	default_resolutions[String("Samsung Galaxy S26+")] = Size2i(1440, 3120);
	default_resolutions[String("Samsung Galaxy S26 Ultra")] = Size2i(1440, 3120);
```

syntaxerror247 commented:
Tested this PR locally ~and I didn't really get the need for this (without display frame)~ I guess device frame can be implemented seperately.

- Can you remove duplicates? There's devices with exact same resolution (iPhone 17 and iPhone 17 Pro, Galaxy S26+ and Galaxy S26 Ultra, etc.)
- How exactly `safe_area` affects the final view? `DisplayServer.get_display_safe_area()` doesn't return those values. I didn't understand how it works (haven't looked at code yet).
- why there's no safearea defined (by default) for Android devices?

AdriaandeJongh commented:
I didn't review thoroughly but some more obvious usability suggestions:
- I would consider setting all resolution preset options to 'disabled' (faded, unclickable) when Stretch to Fit is enabled, thus making the relationship between the embedded window mode and the resolution presets much clearer. Right now the relationship between the two groups is unclear and messy.
- This device list will be high-maintenance, requiring yearly updating. Are we sure about this?

m4gr3d commented:
> * This device list will be high-maintenance, requiring yearly updating. Are we sure about this?

@AdriaandeJongh We touched on this in https://github.com/godotengine/godot/pull/111787/changes#r3581508008; so long as we keep the list small (drop older devices when adding new ones), and have a consistent source for the data, the maintenance should not be an issue. Android and iOS already update at least once a year due to new SDK requirements, we can make this list part of the update.

m4gr3d commented on editor/settings/editor_settings.cpp:
```diff
@@ -1138,6 +1139,57 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
+	default_resolutions[String("iPad Air 13")] = Size2i(2048, 2732);
+	default_resolutions[String("iPad Pro 11")] = Size2i(1668, 2420);
+	default_resolutions[String("iPad Pro 13")] = Size2i(2064, 2752);
+	default_resolutions[String("Samsung Galaxy S26")] = Size2i(1080, 2340);
+	default_resolutions[String("Samsung Galaxy S26+")] = Size2i(1440, 3120);
+	default_resolutions[String("Samsung Galaxy S26 Ultra")] = Size2i(1440, 3120);
+	default_resolutions[String("Google Pixel 9")] = Size2i(1080, 2424);
+	default_resolutions[String("Google Pixel 9 Pro")] = Size2i(1280, 2856);
+	default_resolutions[String("Google Pixel 9 Pro XL")] = Size2i(1344, 2992);
+	default_resolutions[String("OnePlus 13")] = Size2i(1440, 3168);
```
For Android, we may want to stick to the Pixel devices only and only include the latest available.

ajreckof commented:
I did not look at the code or test it yet but here are my usability suggestions:
- I feel like the list is huge and makes it harder to find the one you are looking for, so when you don't actually plan to export to this becomes a huge list taking space for nothing. 
- I feel like this should not appear as long as i don't have an export template that makes sense for it. (ie. if i don't have iOS export, i don't have the iphones resolutions). Either those would automatically detected from export or you would be able to turn them on off manually. Maybe a mix of the two with a on/default/off option with default option detecting export.
- maybe to reduce the size of the list they could be nested inside a sub-menu per platform.
- I would include more brand especially those that can come with a pre-installed open source OS, also we could also add options for desktop -> macos as a really limited number of options. For linux we could have steamDeck
- I think we should directly let EditorExportPlugin add new ones which would reduce the need for adding more as defaults.
